Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialSourceData.cpp

```diff
@@ -188,64 +188,74 @@ namespace AZ
             if (!AZ::RPI::JsonUtils::LoadObjectFromFile(materialTypeSourcePath, materialTypeSourceData))
             {
                 return Failure();
             }
 
             materialTypeSourceData.ResolveUvEnums();
 
-            auto materialTypeAsset =
-                materialTypeSourceData.CreateMaterialTypeAsset(AZ::Uuid::CreateRandom(), materialTypeSourcePath, elevateWarnings);
-            if (!materialTypeAsset.IsSuccess())
-            {
-                return Failure();
-            }
-#else
-            auto materialTypeAsset = AssetUtils::LoadAsset<MaterialTypeAsset>(materialSourceFilePath, m_materialType);
+            const auto materialTypeAsset =
+                materialTypeSourceData.CreateMaterialTypeAsset(materialTypeAssetId.GetValue(), materialTypeSourcePath, elevateWarnings);
             if (!materialTypeAsset.IsSuccess())
             {
                 return Failure();
             }
-#endif
 
-            materialAssetCreator.Begin(assetId, *materialTypeAsset.GetValue().Get(), includeMaterialPropertyNames);
+            AZStd::unordered_set<AZStd::string> dependencies;
+            dependencies.insert(materialSourceFilePath);
+            dependencies.insert(materialTypeSourcePath);
 
             AZStd::vector<MaterialSourceData> parentSourceDataStack;
 
             AZStd::string parentSourceRelPath = m_parentMaterial;
             AZStd::string parentSourceAbsPath = AssetUtils::ResolvePathReference(materialSourceFilePath, parentSourceRelPath);
             while (!parentSourceRelPath.empty())
             {
+                if (dependencies.find(parentSourceAbsPath) != dependencies.end())
+                {
+                    return Failure();
+                }
+
+                dependencies.insert(parentSourceAbsPath);
+
                 MaterialSourceData parentSourceData;
                 if (!AZ::RPI::JsonUtils::LoadObjectFromFile(parentSourceAbsPath, parentSourceData))
                 {
                     return Failure();
                 }
 
                 // Make sure the parent material has the same material type
-                auto materialTypeIdOutcome1 = AssetUtils::MakeAssetId(materialSourceFilePath, m_materialType, 0);
-                auto materialTypeIdOutcome2 = AssetUtils::MakeAssetId(parentSourceAbsPath, parentSourceData.m_materialType, 0);
-                if (!materialTypeIdOutcome1.IsSuccess() || !materialTypeIdOutcome2.IsSuccess() ||
-                    materialTypeIdOutcome1.GetValue() != materialTypeIdOutcome2.GetValue())
+                const auto parentTypeAssetId = AssetUtils::MakeAssetId(parentSourceAbsPath, parentSourceData.m_materialType, 0);
+                if (!parentTypeAssetId || parentTypeAssetId.GetValue() != materialTypeAssetId.GetValue())
                 {
                     AZ_Error("MaterialSourceData", false, "This material and its parent material do not share the same material type.");
```

Comment:
We should probably have separate error messages for parentTypeAssetId being invalid and the material types not matching.